kong/db/errors.lua

```diff
@@ -24,14 +24,16 @@ end
 
 
 local ERRORS            = {
-  INVALID_PRIMARY_KEY   = 1,
-  SCHEMA_VIOLATION      = 2,
-  PRIMARY_KEY_VIOLATION = 3, -- primary key already exists (HTTP 400)
-  FOREIGN_KEY_VIOLATION = 4, -- foreign entity does not exist (HTTP 400)
-  UNIQUE_VIOLATION      = 5, -- unique key already exists (HTTP 409)
-  NOT_FOUND             = 6, -- WHERE clause leads nowhere (HTTP 404)
-  INVALID_OFFSET        = 7, -- page(size, offset) is invalid
-  DATABASE_ERROR        = 8, -- connection refused or DB error (HTTP 500)
+  INVALID_PRIMARY_KEY   =  1,
+  SCHEMA_VIOLATION      =  2,
+  PRIMARY_KEY_VIOLATION =  3, -- primary key already exists (HTTP 400)
+  FOREIGN_KEY_VIOLATION =  4, -- foreign entity does not exist (HTTP 400)
+  UNIQUE_VIOLATION      =  5, -- unique key already exists (HTTP 409)
+  NOT_FOUND             =  6, -- WHERE clause leads nowhere (HTTP 404)
+  INVALID_OFFSET        =  7, -- page(size, offset) is invalid
+  DATABASE_ERROR        =  8, -- connection refused or DB error (HTTP 500)
+  CONFLICTING_INPUT     =  9, -- user-provided data generated a conflict (HTTP 409)
+  INVALID_INPUT         = 10, -- user-provided data is not valid (HTTP 400)
```

Comment:
> This looks conflicting with the above `SCHEMA_VIOLATION` error

I have mixed feelings about reusing the error messages for errors detected via different sources.

On the one hand, for the user perspective I agree that the way bad input parameters were caught shouldn't affect the error message, it's an implementation detail. On the other hand, debugging gets harder when we get reports of `SCHEMA_VIOLATION` from the field and these don't come from the schema library.

I don't think it's the case for certificates, but if something like this starts happening in other parts of the code it could easily trip up plugin authors ("but my schema looks okay!"). But as you said, certs/snis are "abnormal" in that sense, so it's probably fine to make an exception for it in the name of better user-visible consistency.

In that case, the `SCHEMA_VIOLATION` should be formatted in a way consistent with those produced by the schema library (e.g. listing the offending fields, etc).

> I also dislike that this error producer accepts any error message

Definitely agree with this. It shouldn't just forward the full error message, it should construct a message like the other functions in this module.

Reply:
I have removed both my custom errors and used schema violations everywhere. All the errors happen on the context of the `snis` magical pseudo-attribute, so it does not look bad (all errors appear "scoped" by `snis`).